rocketmq-docker/4.2.0/rocketmq-base/scripts/runbroker-customize.sh

```diff
@@ -36,13 +36,22 @@ export CLASSPATH=.:${BASE_DIR}/conf:${CLASSPATH}
 #===========================================================================================
 # JVM Configuration
 #===========================================================================================
-JAVA_OPT="${JAVA_OPT} -server -Xms256m -Xmx256m -Xmn128m"
+# Get the max heap used by a jvm, which used all the ram available to the container.
+MAX_POSSIBLE_HEAP=$(java -XX:+UnlockExperimentalVMOptions -XX:MaxRAMFraction=1 -XshowSettings:vm -version |& awk '/Max\. Heap Size \(Estimated\): [0-9KMG]+/{ print $5}' | gawk -f to_bytes.gawk)
```

Comment:
@huanwei I prefer to use the values calculated from `${MAX_POSSIBLE_HEAP}` as default values so that users can tune their process by override parameter `-xmx`, `-xms`, `-xmn`. Would you mind explaining your consideration and concern?

Reply:
ping @huanwei @vongosling

Reply:
@XiaoZYang Here my consideration is that if we just simply set the max heap used by the container jvm(`${MAX_POSSIBLE_HEAP}`) as default value of `Xms`, `xmx` and `xmn`, there would be few jvm heap left for other process(though currently I have no idea on how many other processes there are for instance in a single namesrv or broker container.).. So I think we need to leave some jvm heap for the other process in the container.. So, in order to keep the containers run smoothly and leave users from pain of performance tunning I set the values as the following (from my limited experience):
```
Xms=$[MAX_POSSIBLE_HEAP/4]
Xmx=$[MAX_POSSIBLE_HEAP/4]
Xmn=$[MAX_POSSIBLE_HEAP/8]
```

Reply:
@XiaoZYang Some supplements for my setting for the default jvm. 
1) Here ${MAX_POSSIBLE_HEAP} also means the total available jvm heap on the Node(on which docker daemon runs) if user don't set the memory resource limit for the namesrv/broker container (especially in Kubernetes scenario there would be many other Pods running on the same Node).
2) for most of user scenarios, like single mode with one namesrv and one broker, they would run on the same docker daemon(or same K8S Pod).So I think it seems make sense totally by default the core rocket(namesrv+broker) use max of 1/2 of the Node's heap..
What do you think? 

Thanks,
Huan Wei

Reply:
That's great! Buf which my suggestion is providing tuning way for expert users. One user can not only run the container with the JVM options calculated by the script but also override the options on his/her own need. So the option (e.g. `-xmx`) here is user given value if provided otherwise default value. I think is more flexible.